engines/scalac/src/main/scala/scala/macros/internal/engines/scalac/semantic/Abstracts.scala

```diff
@@ -163,8 +163,9 @@ trait Abstracts extends scala.macros.semantic.Mirrors { self: Universe =>
     }
 
     implicit class XtensionToGType(tpe: Type) {
-      def toGType: g.Type = tpe match {
+      def toGType(implicit m: Mirror): g.Type = tpe match {
         case gtpt: g.TypeTree => gtpt.tpe
+        case tname: Name => Symbol(tname.value)(m).tpe
         case _ => ???
       }
```

Comment:
I suggest that we represent userland types as corresponding userland trees, e.g. `List[Int]` becomes `Type.Apply(Type.Name(<ListSymbol>), List(Type.Name(<IntSymbol>)))`.

To implement this, `.toType` will have to convert types from compiler's `g.Type` to this representation and `.toGType` will have to do an inverse conversion. We'll also need to call `.toType` in https://github.com/scalamacros/scalamacros/blob/f03bbf3abac159fb1aef420b03088658ee1f808a/plugins/scalac/src/main/scala/scala/macros/internal/plugins/scalac/parser/SyntaxAnalyzer.scala#L306.

Could you take a stab at implementing this?